## Import needed module

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

## Load dataset
Dataset was taken from UCI Machine Learning Repository.
Link : https://archive.ics.uci.edu/ml/datasets/automobile

In [2]:
df_car = pd.read_csv('imports-85.data', names = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
                                                 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels',
                                                 'engine-location', 'wheel-base', 'length', 'width', 
                                                 'height', 'curb-weight', 'engine-type', 'num-of-cylinders', 
                                                 'engine-size', 'fuel-system', 'bore', 'stroke', 
                                                 'compression-ratio', 'horsepower', 'peak-rpm',
                                                 'city-mpg', 'highway-mpg', 'price'])
# display records of all columns
pd.set_option('display.max_columns', None)

df_car.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  length  width  \
0  convertible          rwd           front        88.6   168.8   64.1   
1  convertible          rwd           front        88.6   168.8   64.1   
2    hatchback          rwd           front        94.5   171.2   65.5   
3        sedan          fwd           front        99.8   176.6   66.2   
4        sedan          4wd           front        99.4   176.6   66.4   
5        sedan          fwd           front        99.8   177.3   66.3   
6        sedan          fwd           front       105.8   192.7   71.4   
7        wagon          fwd           front       105.8   192.7   71.4   
8        sedan          fwd           front       105.8   192.7   71.4   
9    hatchback          4wd           front        99.5   178.2   67.9   

   height  curb-weight engine-type num-of-cylinders  engine-size fuel-system  \
0    48.8         2548        dohc             four          130        mpfi   
1    48.8         2548        dohc             four          130        mpfi   
2    52.4         2823        ohcv              six          152        mpfi   
3    54.3         2337         ohc             four          109        mpfi   
4    54.3         2824         ohc             five          136        mpfi   
5    53.1         2507         ohc             five          136        mpfi   
6    55.7         2844         ohc             five          136        mpfi   
7    55.7         2954         ohc             five          136        mpfi   
8    55.9         3086         ohc             five          131        mpfi   
9    52.0         3053         ohc             five          131        mpfi   

   bore stroke  compression-ratio horsepower peak-rpm  city-mpg  highway-mpg  \
0  3.47   2.68                9.0        111     5000        21           27   
1  3.47   2.68                9.0        111     5000        21           27   
2  2.68   3.47                9.0        154     5000        19           26   
3  3.19   3.40               10.0        102     5500        24           30   
4  3.19   3.40                8.0        115     5500        18           22   
5  3.19   3.40                8.5        110     5500        19           25   
6  3.19   3.40                8.5        110     5500        19           25   
7  3.19   3.40                8.5        110     5500        19           25   
8  3.13   3.40                8.3        140     5500        17           20   
9  3.13   3.40                7.0        160     5500        16           22   

   price  
0  13495  
1  16500  
2  16500  
3  13950  
4  17450  
5  15250  
6  17710  
7  18920  
8  23875  
9      ?

## Check if there are missing values

In [3]:
df_car.isnull().any()

symboling            False
normalized-losses    False
make                 False
fuel-type            False
aspiration           False
num-of-doors         False
body-style           False
drive-wheels         False
engine-location      False
wheel-base           False
length               False
width                False
height               False
curb-weight          False
engine-type          False
num-of-cylinders     False
engine-size          False
fuel-system          False
bore                 False
stroke               False
compression-ratio    False
horsepower           False
peak-rpm             False
city-mpg             False
highway-mpg          False
price                False
dtype: bool

As you can see above, each column returns false even though there are mark "?" in some columns. The "?" does not identify as missing values. Therefore, I replace "?" with "NaN" in order that it can be identified as missing values.

In [4]:
df_car.head(50)

symboling normalized-losses         make fuel-type aspiration  \
0           3                 ?  alfa-romero       gas        std   
1           3                 ?  alfa-romero       gas        std   
2           1                 ?  alfa-romero       gas        std   
3           2               164         audi       gas        std   
4           2               164         audi       gas        std   
5           2                 ?         audi       gas        std   
6           1               158         audi       gas        std   
7           1                 ?         audi       gas        std   
8           1               158         audi       gas      turbo   
9           0                 ?         audi       gas      turbo   
10          2               192          bmw       gas        std   
11          0               192          bmw       gas        std   
12          0               188          bmw       gas        std   
13          0               188          bmw       gas        std   
14          1                 ?          bmw       gas        std   
15          0                 ?          bmw       gas        std   
16          0                 ?          bmw       gas        std   
17          0                 ?          bmw       gas        std   
18          2               121    chevrolet       gas        std   
19          1                98    chevrolet       gas        std   
20          0                81    chevrolet       gas        std   
21          1               118        dodge       gas        std   
22          1               118        dodge       gas        std   
23          1               118        dodge       gas      turbo   
24          1               148        dodge       gas        std   
25          1               148        dodge       gas        std   
26          1               148        dodge       gas        std   
27          1               148        dodge       gas      turbo   
28         -1               110        dodge       gas        std   
29          3               145        dodge       gas      turbo   
30          2               137        honda       gas        std   
31          2               137        honda       gas        std   
32          1               101        honda       gas        std   
33          1               101        honda       gas        std   
34          1               101        honda       gas        std   
35          0               110        honda       gas        std   
36          0                78        honda       gas        std   
37          0               106        honda       gas        std   
38          0               106        honda       gas        std   
39          0                85        honda       gas        std   
40          0                85        honda       gas        std   
41          0                85        honda       gas        std   
42          1               107        honda       gas        std   
43          0                 ?        isuzu       gas        std   
44          1                 ?        isuzu       gas        std   
45          0                 ?        isuzu       gas        std   
46          2                 ?        isuzu       gas        std   
47          0               145       jaguar       gas        std   
48          0                 ?       jaguar       gas        std   
49          0                 ?       jaguar       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  length  \
0           two  convertible          rwd           front        88.6   168.8   
1           two  convertible          rwd           front        88.6   168.8   
2           two    hatchback          rwd           front        94.5   171.2   
3          four        sedan          fwd           front        99.8   176.6   
4          four        sedan          4wd           front        99.4   176.6  

## Replace ? with NaN for numeric columns and change the column's type as float

In [5]:
numeric_col = ['symboling', 'normalized-losses', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-size', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

print('----- Count of missing value "?" in numeric columns -----\n')
for col in numeric_col:
    if (df_car[df_car[col]=='?'].shape[0]>0):
        print(col, ':', df_car[df_car[col]=='?'].shape[0])
        df_car[col] = df_car[col].replace('?', np.NaN)
        df_car[col] = df_car[col].astype('float64')
        
df_car.dtypes

----- Count of missing value "?" in numeric columns -----

normalized-losses : 41
bore : 4
stroke : 4
horsepower : 2
peak-rpm : 2
price : 4


C:\Users\levono\Anaconda3\lib\site-packages\pandas\core\ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


symboling              int64
normalized-losses    float64
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

There are 6 columns having "?" with its each amount you can see above and the data types for those columns also changed.

## Re-check for the existing of missing values

In [6]:
df_car.isnull().any()

symboling            False
normalized-losses     True
make                 False
fuel-type            False
aspiration           False
num-of-doors         False
body-style           False
drive-wheels         False
engine-location      False
wheel-base           False
length               False
width                False
height               False
curb-weight          False
engine-type          False
num-of-cylinders     False
engine-size          False
fuel-system          False
bore                  True
stroke                True
compression-ratio    False
horsepower            True
peak-rpm              True
city-mpg             False
highway-mpg          False
price                 True
dtype: bool

Now, you can see that the missing values in those columns were detected.

In [7]:
df_car.head(50)

symboling  normalized-losses         make fuel-type aspiration  \
0           3                NaN  alfa-romero       gas        std   
1           3                NaN  alfa-romero       gas        std   
2           1                NaN  alfa-romero       gas        std   
3           2              164.0         audi       gas        std   
4           2              164.0         audi       gas        std   
5           2                NaN         audi       gas        std   
6           1              158.0         audi       gas        std   
7           1                NaN         audi       gas        std   
8           1              158.0         audi       gas      turbo   
9           0                NaN         audi       gas      turbo   
10          2              192.0          bmw       gas        std   
11          0              192.0          bmw       gas        std   
12          0              188.0          bmw       gas        std   
13          0              188.0          bmw       gas        std   
14          1                NaN          bmw       gas        std   
15          0                NaN          bmw       gas        std   
16          0                NaN          bmw       gas        std   
17          0                NaN          bmw       gas        std   
18          2              121.0    chevrolet       gas        std   
19          1               98.0    chevrolet       gas        std   
20          0               81.0    chevrolet       gas        std   
21          1              118.0        dodge       gas        std   
22          1              118.0        dodge       gas        std   
23          1              118.0        dodge       gas      turbo   
24          1              148.0        dodge       gas        std   
25          1              148.0        dodge       gas        std   
26          1              148.0        dodge       gas        std   
27          1              148.0        dodge       gas      turbo   
28         -1              110.0        dodge       gas        std   
29          3              145.0        dodge       gas      turbo   
30          2              137.0        honda       gas        std   
31          2              137.0        honda       gas        std   
32          1              101.0        honda       gas        std   
33          1              101.0        honda       gas        std   
34          1              101.0        honda       gas        std   
35          0              110.0        honda       gas        std   
36          0               78.0        honda       gas        std   
37          0              106.0        honda       gas        std   
38          0              106.0        honda       gas        std   
39          0               85.0        honda       gas        std   
40          0               85.0        honda       gas        std   
41          0               85.0        honda       gas        std   
42          1              107.0        honda       gas        std   
43          0                NaN        isuzu       gas        std   
44          1                NaN        isuzu       gas        std   
45          0                NaN        isuzu       gas        std   
46          2                NaN        isuzu       gas        std   
47          0              145.0       jaguar       gas        std   
48          0                NaN       jaguar       gas        std   
49          0                NaN       jaguar       gas        std   

   num-of-doors   body-style drive-wheels engine-location  wheel-base  length  \
0           two  convertible          rwd           front        88.6   168.8   
1           two  convertible          rwd           front        88.6   168.8   
2           two    hatchback          rwd           front        94.5   171.2   
3          four        sedan          fwd           front        99.8   176.6   
4          four        sedan

## Show the records which contain NaN in column price

In [8]:
df_car_price = df_car[df_car['price'].isna()]
df_car_price.iloc[:,[2,-1]]

make  price
9       audi    NaN
44     isuzu    NaN
45     isuzu    NaN
129  porsche    NaN

## Show the records which contain NaN in at least 1 column

In [9]:
df_car[df_car.isna().any(axis=1)]

symboling  normalized-losses           make fuel-type aspiration  \
0            3                NaN    alfa-romero       gas        std   
1            3                NaN    alfa-romero       gas        std   
2            1                NaN    alfa-romero       gas        std   
5            2                NaN           audi       gas        std   
7            1                NaN           audi       gas        std   
9            0                NaN           audi       gas      turbo   
14           1                NaN            bmw       gas        std   
15           0                NaN            bmw       gas        std   
16           0                NaN            bmw       gas        std   
17           0                NaN            bmw       gas        std   
43           0                NaN          isuzu       gas        std   
44           1                NaN          isuzu       gas        std   
45           0                NaN          isuzu       gas        std   
46           2                NaN          isuzu       gas        std   
48           0                NaN         jaguar       gas        std   
49           0                NaN         jaguar       gas        std   
55           3              150.0          mazda       gas        std   
56           3              150.0          mazda       gas        std   
57           3              150.0          mazda       gas        std   
58           3              150.0          mazda       gas        std   
63           0                NaN          mazda    diesel        std   
66           0                NaN          mazda    diesel        std   
71          -1                NaN  mercedes-benz       gas        std   
73           0                NaN  mercedes-benz       gas        std   
74           1                NaN  mercedes-benz       gas        std   
75           1                NaN        mercury       gas      turbo   
82           3                NaN     mitsubishi       gas      turbo   
83           3                NaN     mitsubishi       gas      turbo   
84           3                NaN     mitsubishi       gas      turbo   
109          0                NaN         peugot       gas        std   
110          0                NaN         peugot    diesel      turbo   
113          0                NaN         peugot       gas        std   
114          0                NaN         peugot    diesel      turbo   
124          3                NaN       plymouth       gas      turbo   
126          3                NaN        porsche       gas        std   
127          3                NaN        porsche       gas        std   
128          3                NaN        porsche       gas        std   
129          1                NaN        porsche       gas        std   
130          0                NaN        renault       gas        std   
131          2                NaN        renault       gas        std   
181         -1                NaN         toyota       gas        std   
189          3                NaN     volkswagen       gas        std   
191          0                NaN     volkswagen       gas        std   
192          0                NaN     volkswagen    diesel      turbo   
193          0                NaN     volkswagen       gas        std   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  \
0            two  convertible          rwd           front        88.6   
1            two  convertible          rwd           front        88.6   
2            two    hatchback          rwd           front        94.5   
5            two        sedan          fwd           front        99.8   
7           four        wagon          fwd           front       105.8   
9            two    hatchback          4wd           front        99.5   
14          four        sedan          rwd           front       103.5   
15          four        sedan          rwd           f

## Replace NaN in column price with mean based on column make

In [10]:
df_car['price'].fillna(df_car.groupby('make')['price'].transform('mean'), inplace=True)

## Re-check if there is still any missing values in column price

In [11]:
df_car['price'].isna().any()

False

## Show the result of the replacing process

In [12]:
df_car.iloc[[9,44,45,129],[2,-1]]

make         price
9       audi  17859.166667
44     isuzu   8916.500000
45     isuzu   8916.500000
129  porsche  31400.500000

## Show the records which contain NaN in column normalized-losses 

In [13]:
df_car_nl = df_car[df_car['normalized-losses'].isna()]
df_car_nl.iloc[:,[1,2]]

normalized-losses           make
0                  NaN    alfa-romero
1                  NaN    alfa-romero
2                  NaN    alfa-romero
5                  NaN           audi
7                  NaN           audi
9                  NaN           audi
14                 NaN            bmw
15                 NaN            bmw
16                 NaN            bmw
17                 NaN            bmw
43                 NaN          isuzu
44                 NaN          isuzu
45                 NaN          isuzu
46                 NaN          isuzu
48                 NaN         jaguar
49                 NaN         jaguar
63                 NaN          mazda
66                 NaN          mazda
71                 NaN  mercedes-benz
73                 NaN  mercedes-benz
74                 NaN  mercedes-benz
75                 NaN        mercury
82                 NaN     mitsubishi
83                 NaN     mitsubishi
84                 NaN     mitsubishi
109                NaN         peugot
110                NaN         peugot
113                NaN         peugot
114                NaN         peugot
124                NaN       plymouth
126                NaN        porsche
127                NaN        porsche
128                NaN        porsche
129                NaN        porsche
130                NaN        renault
131                NaN        renault
181                NaN         toyota
189                NaN     volkswagen
191                NaN     volkswagen
192                NaN     volkswagen
193                NaN     volkswagen

## Replace the mssing values in column normalized-losses with mean based on column make

In [14]:
df_car['normalized-losses'].fillna(df_car.groupby('make')['normalized-losses'].transform('mean'), inplace=True)

In [15]:
df_car['normalized-losses'].isna().any()

True

All of the alfa-romero, isuzu, mercury, and renault made cars do not have values in its normalized-losses column. So, replacing with mean did not work for them because no single number can be counted to get their mean. The following code shows the containing rows of them.

In [16]:
df_car_nl = df_car[df_car['normalized-losses'].isna()]
df_car_nl.iloc[:,[1,2]]

normalized-losses         make
0                  NaN  alfa-romero
1                  NaN  alfa-romero
2                  NaN  alfa-romero
43                 NaN        isuzu
44                 NaN        isuzu
45                 NaN        isuzu
46                 NaN        isuzu
75                 NaN      mercury
130                NaN      renault
131                NaN      renault

The following code shows the missing values in column bore, stroke, horsepower, and peak-rpm. All of the missing values are going to be removed including the rest of missing values in column normalized-losses.

In [17]:
df_car_nl = df_car[(df_car['bore'].isna())|(df_car['stroke'].isna())|(df_car['horsepower'].isna())|(df_car['peak-rpm'].isna())]
df_car_nl.iloc[:,[2,18,19,21,22]]

make  bore  stroke  horsepower  peak-rpm
55     mazda   NaN     NaN       101.0    6000.0
56     mazda   NaN     NaN       101.0    6000.0
57     mazda   NaN     NaN       101.0    6000.0
58     mazda   NaN     NaN       135.0    6000.0
130  renault  3.46     3.9         NaN       NaN
131  renault  3.46     3.9         NaN       NaN

## Remove all of the rest missing values in the numeric columns

In [18]:
df_car = df_car.dropna(how='any')
df_car.shape

(191, 26)

In [19]:
df_car

symboling  normalized-losses        make fuel-type aspiration  \
3            2         164.000000        audi       gas        std   
4            2         164.000000        audi       gas        std   
5            2         161.000000        audi       gas        std   
6            1         158.000000        audi       gas        std   
7            1         161.000000        audi       gas        std   
8            1         158.000000        audi       gas      turbo   
9            0         161.000000        audi       gas      turbo   
10           2         192.000000         bmw       gas        std   
11           0         192.000000         bmw       gas        std   
12           0         188.000000         bmw       gas        std   
13           0         188.000000         bmw       gas        std   
14           1         190.000000         bmw       gas        std   
15           0         190.000000         bmw       gas        std   
16           0         190.000000         bmw       gas        std   
17           0         190.000000         bmw       gas        std   
18           2         121.000000   chevrolet       gas        std   
19           1          98.000000   chevrolet       gas        std   
20           0          81.000000   chevrolet       gas        std   
21           1         118.000000       dodge       gas        std   
22           1         118.000000       dodge       gas        std   
23           1         118.000000       dodge       gas      turbo   
24           1         148.000000       dodge       gas        std   
25           1         148.000000       dodge       gas        std   
26           1         148.000000       dodge       gas        std   
27           1         148.000000       dodge       gas      turbo   
28          -1         110.000000       dodge       gas        std   
29           3         145.000000       dodge       gas      turbo   
30           2         137.000000       honda       gas        std   
31           2         137.000000       honda       gas        std   
32           1         101.000000       honda       gas        std   
..         ...                ...         ...       ...        ...   
175         -1          65.000000      toyota       gas        std   
176         -1          65.000000      toyota       gas        std   
177         -1          65.000000      toyota       gas        std   
178          3         197.000000      toyota       gas        std   
179          3         197.000000      toyota       gas        std   
180         -1          90.000000      toyota       gas        std   
181         -1         110.290323      toyota       gas        std   
182          2         122.000000  volkswagen    diesel        std   
183          2         122.000000  volkswagen       gas        std   
184          2          94.000000  volkswagen    diesel        std   
185          2          94.000000  volkswagen       gas        std   
186          2          94.000000  volkswagen       gas        std   
187          2          94.000000  volkswagen    diesel      turbo   
188          2          94.000000  volkswagen       gas        std   
189          3         121.250000  volkswagen       gas        std   
190          3         256.000000  volkswagen       gas        std   
191          0         121.250000  volkswagen       gas        std   
192          0         121.250000  volkswagen    diesel      turbo   
193          0         121.250000  volkswagen       gas        std   
194         -2         103.000000       volvo       gas        std   
195         -1          74.000000       volvo       gas        std   
196         -2         103.000000       volvo       gas        std   
197         -1          74.000000       volvo       gas        std   
198         -2         103.000000       volvo       gas      turbo   
199         -1          74.000000       volvo       gas      turbo   
200         -1 

After removing the missing values, so far the dataset have 191 rows and 26 columns remaining.

## Check if there are still '?' in entire dataset

In [20]:
df_car[(df_car == '?').any(axis = 1)]

symboling  normalized-losses   make fuel-type aspiration num-of-doors  \
27          1         148.000000  dodge       gas      turbo            ?   
63          0         123.933333  mazda    diesel        std            ?   

   body-style drive-wheels engine-location  wheel-base  length  width  height  \
27      sedan          fwd           front        93.7   157.3   63.8    50.6   
63      sedan          fwd           front        98.8   177.8   66.5    55.5   

    curb-weight engine-type num-of-cylinders  engine-size fuel-system  bore  \
27         2191         ohc             four           98        mpfi  3.03   
63         2443         ohc             four          122         idi  3.39   

    stroke  compression-ratio  horsepower  peak-rpm  city-mpg  highway-mpg  \
27    3.39                7.6       102.0    5500.0        24           30   
63    3.39               22.7        64.0    4650.0        36           42   

      price  
27   8558.0  
63  10795.0

## Remove the rows that still contain '?'

In [21]:
df_car = df_car.drop([27,63], axis=0)

In [22]:
df_car

symboling  normalized-losses        make fuel-type aspiration  \
3            2         164.000000        audi       gas        std   
4            2         164.000000        audi       gas        std   
5            2         161.000000        audi       gas        std   
6            1         158.000000        audi       gas        std   
7            1         161.000000        audi       gas        std   
8            1         158.000000        audi       gas      turbo   
9            0         161.000000        audi       gas      turbo   
10           2         192.000000         bmw       gas        std   
11           0         192.000000         bmw       gas        std   
12           0         188.000000         bmw       gas        std   
13           0         188.000000         bmw       gas        std   
14           1         190.000000         bmw       gas        std   
15           0         190.000000         bmw       gas        std   
16           0         190.000000         bmw       gas        std   
17           0         190.000000         bmw       gas        std   
18           2         121.000000   chevrolet       gas        std   
19           1          98.000000   chevrolet       gas        std   
20           0          81.000000   chevrolet       gas        std   
21           1         118.000000       dodge       gas        std   
22           1         118.000000       dodge       gas        std   
23           1         118.000000       dodge       gas      turbo   
24           1         148.000000       dodge       gas        std   
25           1         148.000000       dodge       gas        std   
26           1         148.000000       dodge       gas        std   
28          -1         110.000000       dodge       gas        std   
29           3         145.000000       dodge       gas      turbo   
30           2         137.000000       honda       gas        std   
31           2         137.000000       honda       gas        std   
32           1         101.000000       honda       gas        std   
33           1         101.000000       honda       gas        std   
..         ...                ...         ...       ...        ...   
175         -1          65.000000      toyota       gas        std   
176         -1          65.000000      toyota       gas        std   
177         -1          65.000000      toyota       gas        std   
178          3         197.000000      toyota       gas        std   
179          3         197.000000      toyota       gas        std   
180         -1          90.000000      toyota       gas        std   
181         -1         110.290323      toyota       gas        std   
182          2         122.000000  volkswagen    diesel        std   
183          2         122.000000  volkswagen       gas        std   
184          2          94.000000  volkswagen    diesel        std   
185          2          94.000000  volkswagen       gas        std   
186          2          94.000000  volkswagen       gas        std   
187          2          94.000000  volkswagen    diesel      turbo   
188          2          94.000000  volkswagen       gas        std   
189          3         121.250000  volkswagen       gas        std   
190          3         256.000000  volkswagen       gas        std   
191          0         121.250000  volkswagen       gas        std   
192          0         121.250000  volkswagen    diesel      turbo   
193          0         121.250000  volkswagen       gas        std   
194         -2         103.000000       volvo       gas        std   
195         -1          74.000000       volvo       gas        std   
196         -2         103.000000       volvo       gas        std   
197         -1          74.000000       volvo       gas        std   
198         -2         103.000000       volvo       gas      turbo   
199         -1          74.000000       volvo       gas      turbo   
200         -1 

After removing all of rows that contain '?', now there are 189 rows and 26 columns remaining that are going to be used for the next process.

## Convert column num-of-doors and num-of-cylinders to numeric

First, check the unique values in columns num-of-doors and num-of-cylinders. Next, create functions for converting the values in those columns to numeric. This process will be done in order to make the dataset easy to understand.

In [23]:
df_car['num-of-doors'].unique()

array(['four', 'two'], dtype=object)

In [24]:
df_car['num-of-cylinders'].unique()

array(['four', 'five', 'six', 'three', 'twelve', 'eight'], dtype=object)

In [25]:
def convert_num_of_door(door):
    if (door == 'two'):
        return 2
    if (door == 'four'):
        return 4
    return 0
    
def convert_num_of_cylinder(cylinder):
    cyl = {'three': 3, 'four': 4, 'five': 5, 'six': 6, 'eight': 8, 'twelve': 12}
    if (cylinder in cyl.keys()):
        return cyl[cylinder]
    return 0

In [26]:
df_car['num-of-doors'] = df_car['num-of-doors'].apply(lambda x: convert_num_of_door(x))
df_car['num-of-cylinders'] = df_car['num-of-cylinders'].apply(lambda x: convert_num_of_cylinder(x))

In [27]:
df_car.head(10)

symboling  normalized-losses  make fuel-type aspiration  num-of-doors  \
3           2              164.0  audi       gas        std             4   
4           2              164.0  audi       gas        std             4   
5           2              161.0  audi       gas        std             2   
6           1              158.0  audi       gas        std             4   
7           1              161.0  audi       gas        std             4   
8           1              158.0  audi       gas      turbo             4   
9           0              161.0  audi       gas      turbo             2   
10          2              192.0   bmw       gas        std             2   
11          0              192.0   bmw       gas        std             4   
12          0              188.0   bmw       gas        std             2   

   body-style drive-wheels engine-location  wheel-base  length  width  height  \
3       sedan          fwd           front        99.8   176.6   66.2    54.3   
4       sedan          4wd           front        99.4   176.6   66.4    54.3   
5       sedan          fwd           front        99.8   177.3   66.3    53.1   
6       sedan          fwd           front       105.8   192.7   71.4    55.7   
7       wagon          fwd           front       105.8   192.7   71.4    55.7   
8       sedan          fwd           front       105.8   192.7   71.4    55.9   
9   hatchback          4wd           front        99.5   178.2   67.9    52.0   
10      sedan          rwd           front       101.2   176.8   64.8    54.3   
11      sedan          rwd           front       101.2   176.8   64.8    54.3   
12      sedan          rwd           front       101.2   176.8   64.8    54.3   

    curb-weight engine-type  num-of-cylinders  engine-size fuel-system  bore  \
3          2337         ohc                 4          109        mpfi  3.19   
4          2824         ohc                 5          136        mpfi  3.19   
5          2507         ohc                 5          136        mpfi  3.19   
6          2844         ohc                 5          136        mpfi  3.19   
7          2954         ohc                 5          136        mpfi  3.19   
8          3086         ohc                 5          131        mpfi  3.13   
9          3053         ohc                 5          131        mpfi  3.13   
10         2395         ohc                 4          108        mpfi  3.50   
11         2395         ohc                 4          108        mpfi  3.50   
12         2710         ohc                 6          164        mpfi  3.31   

    stroke  compression-ratio  horsepower  peak-rpm  city-mpg  highway-mpg  \
3     3.40               10.0       102.0    5500.0        24           30   
4     3.40                8.0       115.0    5500.0        18           22   
5     3.40                8.5       110.0    5500.0        19           25   
6     3.40                8.5       110.0    5500.0        19           25   
7     3.40                8.5       110.0    5500.0        19           25   
8     3.40                8.3       140.0    5500.0        17           20   
9     3.40                7.0       160.0    5500.0        16           22   
10    2.80                8.8       101.0    5800.0        23           29   
11    2.80                8.8       101.0    5800.0        23           29   
12    3.19                9.0       121.0    4250.0        21           28   

           price  
3   13950.000000  
4   17450.000000  
5   15250.000000  
6   17710.000000  
7   18920.000000  
8   23875.000000  
9   17859.166667  
10  16430.000000  
11  16925.000000  
12  20970.000000

As you can see below, after converting process, data types in columns num-of-doors and num-of-cylinders change into integer.

In [28]:
df_car.dtypes

symboling              int64
normalized-losses    float64
make                  object
fuel-type             object
aspiration            object
num-of-doors           int64
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders       int64
engine-size            int64
fuel-system           object
bore                 float64
stroke               float64
compression-ratio    float64
horsepower           float64
peak-rpm             float64
city-mpg               int64
highway-mpg            int64
price                float64
dtype: object

## Convert column drive-wheels in order to easier to understand

The process is similar to the previous converting process. First, check the unique values in the column drive-wheels, then make a function for converting process as you can see below.

In [29]:
df_car['drive-wheels'].unique()

array(['fwd', '4wd', 'rwd'], dtype=object)

In [30]:
def convert_drive_wheel(wheels):
    if (wheels == 'fwd'):
        return 'Front Wheel Drive'
    if (wheels == '4wd'):
        return 'Four Wheel Drive'
    if (wheels == 'rwd'):
        return 'Rear Wheel Drive'
    return 0

In [31]:
df_car['drive-wheels'] = df_car['drive-wheels'].apply(lambda x: convert_drive_wheel(x))

In [32]:
df_car.head(10)

symboling  normalized-losses  make fuel-type aspiration  num-of-doors  \
3           2              164.0  audi       gas        std             4   
4           2              164.0  audi       gas        std             4   
5           2              161.0  audi       gas        std             2   
6           1              158.0  audi       gas        std             4   
7           1              161.0  audi       gas        std             4   
8           1              158.0  audi       gas      turbo             4   
9           0              161.0  audi       gas      turbo             2   
10          2              192.0   bmw       gas        std             2   
11          0              192.0   bmw       gas        std             4   
12          0              188.0   bmw       gas        std             2   

   body-style       drive-wheels engine-location  wheel-base  length  width  \
3       sedan  Front Wheel Drive           front        99.8   176.6   66.2   
4       sedan   Four Wheel Drive           front        99.4   176.6   66.4   
5       sedan  Front Wheel Drive           front        99.8   177.3   66.3   
6       sedan  Front Wheel Drive           front       105.8   192.7   71.4   
7       wagon  Front Wheel Drive           front       105.8   192.7   71.4   
8       sedan  Front Wheel Drive           front       105.8   192.7   71.4   
9   hatchback   Four Wheel Drive           front        99.5   178.2   67.9   
10      sedan   Rear Wheel Drive           front       101.2   176.8   64.8   
11      sedan   Rear Wheel Drive           front       101.2   176.8   64.8   
12      sedan   Rear Wheel Drive           front       101.2   176.8   64.8   

    height  curb-weight engine-type  num-of-cylinders  engine-size  \
3     54.3         2337         ohc                 4          109   
4     54.3         2824         ohc                 5          136   
5     53.1         2507         ohc                 5          136   
6     55.7         2844         ohc                 5          136   
7     55.7         2954         ohc                 5          136   
8     55.9         3086         ohc                 5          131   
9     52.0         3053         ohc                 5          131   
10    54.3         2395         ohc                 4          108   
11    54.3         2395         ohc                 4          108   
12    54.3         2710         ohc                 6          164   

   fuel-system  bore  stroke  compression-ratio  horsepower  peak-rpm  \
3         mpfi  3.19    3.40               10.0       102.0    5500.0   
4         mpfi  3.19    3.40                8.0       115.0    5500.0   
5         mpfi  3.19    3.40                8.5       110.0    5500.0   
6         mpfi  3.19    3.40                8.5       110.0    5500.0   
7         mpfi  3.19    3.40                8.5       110.0    5500.0   
8         mpfi  3.13    3.40                8.3       140.0    5500.0   
9         mpfi  3.13    3.40                7.0       160.0    5500.0   
10        mpfi  3.50    2.80                8.8       101.0    5800.0   
11        mpfi  3.50    2.80                8.8       101.0    5800.0   
12        mpfi  3.31    3.19                9.0       121.0    4250.0   

    city-mpg  highway-mpg         price  
3         24           30  13950.000000  
4         18           22  17450.000000  
5         19           25  15250.000000  
6         19           25  17710.000000  
7         19           25  18920.000000  
8         17           20  23875.000000  
9         16           22  17859.166667  
10        23           29  16430.000000  
11        23           29  16925.000000  
12        21           28  20970.000000

## Convert MPG to L/100km in order to easier to understand
Liter per 100 km (L/100km) is more common measure of car's fuel consumption than Miles Per Gallon (MPG). So, to make easier to understand, I am going to convert MPG to L/100km.

In [33]:
# convert MPG to L/100km for column city-mpg
df_car['city-mpg'] = 235.215/df_car['city-mpg']
# rename the column city-mpg to city-L/100km
df_car.rename(columns = {'city-mpg':'city-L/100km'}, inplace=True)

# convert MPG to L/101km for column highway-mpg
df_car['highway-mpg'] = 235.215/df_car['highway-mpg']
# rename the column highway-mpg to highway-L/100km
df_car.rename(columns = {'highway-mpg':'highway-L/100km'}, inplace=True)

In [34]:
df_car

symboling  normalized-losses        make fuel-type aspiration  \
3            2         164.000000        audi       gas        std   
4            2         164.000000        audi       gas        std   
5            2         161.000000        audi       gas        std   
6            1         158.000000        audi       gas        std   
7            1         161.000000        audi       gas        std   
8            1         158.000000        audi       gas      turbo   
9            0         161.000000        audi       gas      turbo   
10           2         192.000000         bmw       gas        std   
11           0         192.000000         bmw       gas        std   
12           0         188.000000         bmw       gas        std   
13           0         188.000000         bmw       gas        std   
14           1         190.000000         bmw       gas        std   
15           0         190.000000         bmw       gas        std   
16           0         190.000000         bmw       gas        std   
17           0         190.000000         bmw       gas        std   
18           2         121.000000   chevrolet       gas        std   
19           1          98.000000   chevrolet       gas        std   
20           0          81.000000   chevrolet       gas        std   
21           1         118.000000       dodge       gas        std   
22           1         118.000000       dodge       gas        std   
23           1         118.000000       dodge       gas      turbo   
24           1         148.000000       dodge       gas        std   
25           1         148.000000       dodge       gas        std   
26           1         148.000000       dodge       gas        std   
28          -1         110.000000       dodge       gas        std   
29           3         145.000000       dodge       gas      turbo   
30           2         137.000000       honda       gas        std   
31           2         137.000000       honda       gas        std   
32           1         101.000000       honda       gas        std   
33           1         101.000000       honda       gas        std   
..         ...                ...         ...       ...        ...   
175         -1          65.000000      toyota       gas        std   
176         -1          65.000000      toyota       gas        std   
177         -1          65.000000      toyota       gas        std   
178          3         197.000000      toyota       gas        std   
179          3         197.000000      toyota       gas        std   
180         -1          90.000000      toyota       gas        std   
181         -1         110.290323      toyota       gas        std   
182          2         122.000000  volkswagen    diesel        std   
183          2         122.000000  volkswagen       gas        std   
184          2          94.000000  volkswagen    diesel        std   
185          2          94.000000  volkswagen       gas        std   
186          2          94.000000  volkswagen       gas        std   
187          2          94.000000  volkswagen    diesel      turbo   
188          2          94.000000  volkswagen       gas        std   
189          3         121.250000  volkswagen       gas        std   
190          3         256.000000  volkswagen       gas        std   
191          0         121.250000  volkswagen       gas        std   
192          0         121.250000  volkswagen    diesel      turbo   
193          0         121.250000  volkswagen       gas        std   
194         -2         103.000000       volvo       gas        std   
195         -1          74.000000       volvo       gas        std   
196         -2         103.000000       volvo       gas        std   
197         -1          74.000000       volvo       gas        std   
198         -2         103.000000       volvo       gas      turbo   
199         -1          74.000000       volvo       gas      turbo   
200         -1 

## Binning the price into low, medium, and high
Binning can increase the accuracy in modeling and to have better understanding of the data. I divide the price into 3 categories; high, medium, and low.

In [35]:
binwidth = int((max(df_car['price'])-min(df_car['price']))/3)

bins = range(int(min(df_car['price'])), int(max(df_car['price'])), binwidth)

df_car['price-binned'] = pd.cut(df_car['price'], bins, labels=['Low', 'Medium', 'High'])

In [36]:
df_car.loc[:,['price','price-binned']].sample(10)

price price-binned
200  16845.0          Low
66   18344.0          Low
116  17950.0          Low
47   32250.0         High
148   8013.0          Low
79    7689.0          Low
61   10595.0          Low
132  11850.0          Low
74   45400.0          NaN
157   7198.0          Low

## Check if there are missing values after binning process

In [37]:
df_car[(df_car.isna()).any(axis = 1)]

symboling  normalized-losses           make fuel-type aspiration  \
74           1              102.8  mercedes-benz       gas        std   
138          2               83.0         subaru       gas        std   

     num-of-doors body-style       drive-wheels engine-location  wheel-base  \
74              2    hardtop   Rear Wheel Drive           front       112.0   
138             2  hatchback  Front Wheel Drive           front        93.7   

     length  width  height  curb-weight engine-type  num-of-cylinders  \
74    199.2   72.0    55.4         3715        ohcv                 8   
138   156.9   63.4    53.7         2050        ohcf                 4   

     engine-size fuel-system  bore  stroke  compression-ratio  horsepower  \
74           304        mpfi  3.80    3.35                8.0       184.0   
138           97        2bbl  3.62    2.36                9.0        69.0   

     peak-rpm  city-L/100km  highway-L/100km    price price-binned  
74     4500.0     16.801071        14.700938  45400.0          NaN  
138    4900.0      7.587581         6.533750   5118.0          NaN

As we can see above, minimum and maximum value in the dataset unexpectedly did not automatically generate the categories. So, I will label them manually, Low for minimum value and High for maximum value.

In [38]:
df_car.loc[74, 'price-binned'] = 'High'
df_car.loc[138, 'price-binned'] = 'Low'

In [39]:
df_car.loc[[74,138], :]

symboling  normalized-losses           make fuel-type aspiration  \
74           1              102.8  mercedes-benz       gas        std   
138          2               83.0         subaru       gas        std   

     num-of-doors body-style       drive-wheels engine-location  wheel-base  \
74              2    hardtop   Rear Wheel Drive           front       112.0   
138             2  hatchback  Front Wheel Drive           front        93.7   

     length  width  height  curb-weight engine-type  num-of-cylinders  \
74    199.2   72.0    55.4         3715        ohcv                 8   
138   156.9   63.4    53.7         2050        ohcf                 4   

     engine-size fuel-system  bore  stroke  compression-ratio  horsepower  \
74           304        mpfi  3.80    3.35                8.0       184.0   
138           97        2bbl  3.62    2.36                9.0        69.0   

     peak-rpm  city-L/100km  highway-L/100km    price price-binned  
74     4500.0     16.801071        14.700938  45400.0         High  
138    4900.0      7.587581         6.533750   5118.0          Low

In [40]:
df_car.isna().any()

symboling            False
normalized-losses    False
make                 False
fuel-type            False
aspiration           False
num-of-doors         False
body-style           False
drive-wheels         False
engine-location      False
wheel-base           False
length               False
width                False
height               False
curb-weight          False
engine-type          False
num-of-cylinders     False
engine-size          False
fuel-system          False
bore                 False
stroke               False
compression-ratio    False
horsepower           False
peak-rpm             False
city-L/100km         False
highway-L/100km      False
price                False
price-binned         False
dtype: bool

Now, each column does already not have any missing values.

## Store the preprocessed dataset to csv

In [45]:
# save to new csv file
df_car.to_csv('automobile_dataset_preprocessing.csv')